src/fsharp/FSharp.Core/local.fs

```diff
@@ -85,6 +85,42 @@ module internal List =
                 cons <- cons2
             setFreshConsTail cons []
             res
+      
+    let inline groupBy (comparer:IEqualityComparer<'SafeKey>) (keyf:'T->'SafeKey) (getKey:'SafeKey->'Key) (list: 'T list) =
+        let dict = Dictionary<_, _ list []> comparer
+
+        // Build the groupings
+        let rec loop list =
+            match list with
+            | v :: t -> 
+                let safeKey = keyf v
+                match dict.TryGetValue(safeKey) with
+                | true, prev -> 
+                       let cons2 = freshConsNoTail v
+                       setFreshConsTail prev.[1] cons2
+                       prev.[1] <- cons2
+                | _ -> let res = freshConsNoTail v
+                       dict.[safeKey] <- [|res; res |] // First index stores the result list; second index is the most recent cons.
```

Comment:
Since this dictionary always holds a 2-element array it seems it would be better to use a specific new type for this?  An array of two elements is I think 2 words bigger than an object containing two values (one for the length, one for the element type)